src/de.rs

```diff
@@ -1358,3 +1346,78 @@ where
         self.map.next_value_seed(seed)
     }
 }
+
+/// Deserialize a DAG-CBOR encoded CID.
+///
+/// This is without the CBOR tag information. It is only the CBOR byte string identifier (major
+/// type 2), the number of bytes, and a null byte prefixed CID.
+///
+/// The reason for not including the CBOR tag information is the [`Value`] implementation. That one
+/// starts to parse the bytes, before we could interfere. If the data only includes a CID, we are
+/// parsing over the tag to determine whether it is a CID or not and go from there.
+struct CidDeserializer<'a, R>(&'a mut Deserializer<R>);
+
+//impl<'de, 'a: 'de, R> de::Deserializer<'de> for &'a mut CidDeserializer<'a, R>
+impl<'de, 'a, R> de::Deserializer<'de> for &'a mut CidDeserializer<'a, R>
+where
+    R: Read<'de>,
+{
+    type Error = Error;
+
+    fn deserialize_any<V: de::Visitor<'de>>(self, _visitor: V) -> Result<V::Value> {
+        Err(Error::message("Only bytes can be deserialized into a CID"))
+    }
+
+    #[inline]
+    fn deserialize_bytes<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
+        // Match on the major type, it must be a byte string (major type 2)
+        let len = match self.0.parse_u8()? {
+            byte @ 0x40..=0x57 => usize::try_from(byte - 0x40)
+                .map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?,
+            0x58 => {
+                let len = self.0.parse_u8()?;
+                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
+            }
+            0x59 => {
+                let len = self.0.parse_u16()?;
+                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
+            }
+            0x5a => {
+                let len = self.0.parse_u32()?;
+                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
+            }
+            0x5b => {
+                let len = self.0.parse_u64()?;
+                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
+            }
+            _ => Err(self.0.error(ErrorCode::UnexpectedCode))?,
+        };
+
+        match self.0.read.read(len)? {
```

Comment:
what if length is 0? I believe the code below will panic.

Reply:
Good catch!